crates/huub/src/constraints/int_value_precede.rs

```diff
@@ -0,0 +1,1144 @@
+}
+
+impl IntSeqPrecedeChainBounds {
+	/// 3 cases:
+	///
+	/// - Current lower bound is above k - This is the value that required the
+	///   earlier lower bound.
+	/// - k is in the domain of var i - Go one step up and to the next variable.
+	/// - k is not in the domain of var i - i can be anything else, go to next
+	///   variable.
+	fn ex_l<P: PropagationActions>(&self, actions: &mut P, i: usize, k: IntVal) -> Vec<BoolView> {
+		if actions.get_int_lower_bound(self.vars[i]) > k {
+			return vec![actions.get_int_lit(self.vars[i], IntLitMeaning::GreaterEq(k + 1))];
+		}
+		if actions.check_int_in_domain(self.vars[i], k) {
+			return self.ex_l(actions, i + 1, k + 1);
+		}
+		let mut v = self.ex_l(actions, i + 1, k);
+		v.push(actions.get_int_lit(self.vars[i], IntLitMeaning::NotEq(k)));
+		v
+	}
+
+	/// Lower bound explanation: Could not have this value earlier (=upper bound
+	/// explanation) and some later value requires the lower bound.
+	fn explain_lower<P: PropagationActions>(
+		&self,
+		actions: &mut P,
+		i: usize,
+		k: IntVal,
+	) -> Vec<BoolView> {
+		let mut v = self.ex_l(actions, i + 1, k);
+		v.extend(self.explain_upper(actions, i, k));
+		v
+	}
+
+	/// Upper bound explanation: All previous elements are smaller.
+	fn explain_upper<P: PropagationActions>(
+		&self,
+		actions: &mut P,
+		i: usize,
+		k: IntVal,
+	) -> Vec<BoolView> {
+		self.vars
+			.iter()
+			.take(i)
+			.map(|&v| actions.get_int_lit(v, IntLitMeaning::Less(k)))
+			.collect()
+	}
+
+	/// Get the latest occurrence of k, or the maximum variable index if there is
+	/// no latest.
+	fn get_upper_limit<P: PropagationActions>(&self, actions: &mut P, k: usize) -> IntVal {
+		min(
+			actions.get_trailed_int(self.last[k]),
+			self.vars.len() as IntVal - 1,
+		)
+	}
+
+	/// Create a new [`IntSeqPrecedeChainBounds`] propagator and post it in the
+	/// solver.
+	pub fn new_in<P: PropagatorInitActions + ?Sized>(solver: &mut P, vars: Vec<IntView>) {
+		let n = vars.len();
+		let ub = vars.iter().fold(0, |u, &item| {
+			if solver.get_int_upper_bound(item) > u {
+				u + 1
+			} else {
+				u
+			}
+		});
+
+		let first = (0..=ub).map(|_| solver.new_trailed_int(0)).collect();
+		let last = (0..=ub)
+			.map(|i| {
+				if i == 0 {
+					solver.new_trailed_int(IntVal::MIN)
+				} else {
+					solver.new_trailed_int(IntVal::MAX)
+				}
+			})
+			.collect();
+		let first_val = (0..n).map(|_| solver.new_trailed_int(0)).collect();
+		let max_last = solver.new_trailed_int(0);
+
+		let prop = solver.add_propagator(
+			Box::new(Self {
+				vars: vars.clone(),
+				initialized: false,
+				first,
+				last,
+				first_val,
+				max_last,
+			}),
+			PriorityLevel::Low,
+		);
+
+		for v in vars {
+			solver.enqueue_on_int_change(prop, v, IntPropCond::Domain);
+		}
+		solver.enqueue_now(prop);
+	}
+
+	/// Do a full propagation run, requires checking all variables in both
+	/// directions.
+	fn propagate_full<P: PropagationActions>(&mut self, actions: &mut P) -> Result<(), Conflict> {
```

Comment:
This is only done once, maybe rename it to be `initial_propagation` or something.